### PR TITLE
[CSApply] Avoid shortcutting argument conversion when parameter has a…

### PR DESCRIPTION
**Pull request by xedin**
…n external property wrapper

The check to see whether argument matches the parameter exactly causes two problems: 
prevents projected value initialized injection; and, if there are multiple parameters with property 
wrappers, would apply incorrect wrapper to other locations because the wrapper application
index wasn't incremented.

Resolves: rdar://140282980

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Review by DougGregor — APPROVED** [review]
Looks great, thank you!

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -6288,9 +6288,17 @@ ArgumentList *ExprRewriter::coerceCallArguments(
...
+          paramInfo.hasExternalPropertyWrapper(paramIdx))
+        return false;
+
+      return argType->isEqual(paramType);
+    };
+
     auto paramType = param.getOldType();
-    if (argType->isEqual(paramType) && !shouldInjectWrappedValuePlaceholder) {
+
+    if (canShortcutConversion(argType, paramType)) {
```
Comment (hamishknight) [inline lib/Sema/CSApply.cpp:6301]: I do wonder if could do away with this short-circuiting altogether, seems like too easy of a footgun and I'm not sure it's buying us much

Comment (xedin) [reply lib/Sema/CSApply.cpp:6301]: Yeah, that's what I was thinking as well, I just think it might be worth doing that separately...

Comment (hamishknight) [reply lib/Sema/CSApply.cpp:6301]: Oh yeah definitely, doesn't have to be this PR

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**